### ISSUE TITLE
Replace Py_INCREF()/Py_XINCREF() usage with Py_NewRef()/Py_XNewRef()

### ISSUE THREAD

**Issue by vstinner** [python/cpython#99300]
I added Py_NewRef() and Py_XNewRef() to Python 3.10 C API. IMO using them make to code easier to read and make the code looks "more correct". Examples:

(A) Assign + INCREF:

```diff
-                result = Py_False;
-                Py_INCREF(result);
+                result = Py_NewRef(Py_False);
```

(B) INCREF + assign:

```diff
-        Py_INCREF(last);
-        self->last = last;
+        self->last = Py_NewRef(last);
```

(C) INCREF + return:

```diff
-    Py_XINCREF(result);
-    return result;
+    return Py_XNewRef(result);
```

While technically, Py_INCREF() and Py_XINCREF() modify the object in-place (increment their reference counter), for me Py_NewRef() makes me sense: it creates "a new reference".

The example (A) is weird: it assigns a variable to something, and only later creates a new reference. For me, the syntax with Py_NewRef() makes more sense.

Examples (B) and (C) are shorter with Py_NewRef(), and again, IMO makes more sense and are more readable.

<!-- gh-pr-number: gh-99302 -->
* PR: gh-99302
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99317 -->
* PR: gh-99317
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99318 -->
* PR: gh-99318
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99330 -->
* PR: gh-99330
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99332 -->
* PR: gh-99332
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99333 -->
* PR: gh-99333
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99335 -->
* PR: gh-99335
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99336 -->
* PR: gh-99336
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99351 -->
* PR: gh-99351
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99354 -->
* PR: gh-99354
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99436 -->
* PR: gh-99436
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99438 -->
* PR: gh-99438
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99439 -->
* PR: gh-99439
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99440 -->
* PR: gh-99440
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99441 -->
* PR: gh-99441
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99465 -->
* PR: gh-99465
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99466 -->
* PR: gh-99466
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99467 -->
* PR: gh-99467
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99468 -->
* PR: gh-99468
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99469 -->
* PR: gh-99469
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99473 -->
* PR: gh-99473
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99479 -->
* PR: gh-99479
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99480 -->
* PR: gh-99480
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99499 -->
* PR: gh-99499
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99513 -->
* PR: gh-99513
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99530 -->
* PR: gh-99530
<!-- /gh-pr-number -->


<!-- gh-pr-number: gh-99696 -->
* PR: gh-99696
<!-- /gh-pr-number -->

**Comment by gvanrossum** [comment]
I like the new idiom but not so much that I think it was worth all the code churn. (I would never have noticed you weren't doing this if it hadn't been for some merge conflicts in bytecodes.c, so you can ignore this comment. :-)

**Comment by vstinner** [comment]
> I like the new idiom but not so much that I think it was worth all the code churn. (I would never have noticed you weren't doing this if it hadn't been for some merge conflicts in bytecodes.c, so you can ignore this comment. :-)

Sorry for the conflicts :-( I know that it's unpleasant. I tried to group changes related to ceval.c in a single commit to not require people have to fix conflicts multiple times: it's a single shot.

If you have troubles to fix your merge conflicts, I can help you to fix them. I have a tool to replace Py_INCREF with Py_NewRef().

Since I got [positive feedbacks](https://github.com/python/cpython/pull/99441#pullrequestreview-1178280031) on other changes, I continue the work on this issue. The last big piece is the Modules/ directory. Then there are only a few remaining files which need to be changed.

FYI I started to collect some patterns which look suspicious and I plan to investigate them. It seems like I found a few reference leaks, and replacing Py_INCREF with Py_NewRef helped me to identify these.

I'm also considering to replace some "DECREF; INCREF; assign" patterns with Py_SETREF() when applicable. But I will open separated issues for these.
